### PR TITLE
CollectionElement and Autoupdate cleanups to help mypy

### PR DESCRIPTION
**Pull request by ostcar**
* change get_restricted_data and get_projector_data to always use a list
* Add typings to all get_restricted_data and get_projector_data methods
* Replace CollectionElementList with a real list
* Fixed arguments of inform_deleted_data

Should not change anything

### CODE REVIEW

**Comment by ostcar** [comment]
* Refactored the CollectionElement cache
* Run tests with cache enabled (using fakeredis)

**Comment by ostcar** [comment]
@normanjaeckel all done. Please review

File: openslides/utils/access_permissions.py [hunk 1]
```diff
@@ -40,35 +37,30 @@ def get_full_data(self, instance: Model) -> Dict[str, Any]:
...
-            data = None
-        return data
+        return full_data if self.check_permissions(user) else []
 
-    def get_projector_data(self, container: Container) -> RestrictedData:
+    def get_projector_data(self, full_data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
         """
-        Returns the serialized data for the projector. Returns None if the
-        user has no access to this specific data. Returns reduced data if
+        Returns the serialized data for the projector. Returns an empfty list if
```
Comment (normanjaeckel) [inline openslides/utils/access_permissions.py]: typo `empfty`

File: openslides/utils/access_permissions.py [hunk 2]
```diff
@@ -40,35 +37,30 @@ def get_full_data(self, instance: Model) -> Dict[str, Any]:
...
         Returns the restricted serialized data for the instance prepared
         for the user.
 
-        The argument container should be a CollectionElement or a
-        Collection. The type of the return value is a dictionary or a list
-        according to the given type (or None). Returns None or an empty
-        list if the user has no read access. Returns reduced data if the
-        user has limited access. Default: Returns full data if the user has
-        read access to model instances.
+        The argument full_data has to be a list of full_data dicts. Like thes are
```
Comment (normanjaeckel) [inline openslides/utils/access_permissions.py]: typo `thes`

File: openslides/utils/autoupdate.py [hunk 3]
```diff
@@ -231,14 +214,20 @@ def ws_add_projector(message: Any, projector_id: int) -> None:
...
 
             # Collect all elements that are on the projector.
-            output = []
+            output = []  # type: List[AutoupdateFormat]
             for requirement in projector.get_all_requirements():
                 required_collection_element = CollectionElement.from_instance(requirement)
                 output.append(required_collection_element.as_autoupdate_for_projector())
 
             # Collect all config elements.
             collection = Collection(config.get_collection_string())
```
Comment (normanjaeckel) [inline openslides/utils/autoupdate.py]: Please rename `collection` to `config_collection` or something similar.

File: openslides/utils/cache.py [hunk 4]
```diff
@@ -203,9 +204,152 @@ def save_data(self, data: UserCacheDataType) -> None:
...
+        # Clear the cache for collection
+        pipe.delete(self.get_cache_key(collection_string))
+
+        # Save all elements
+        from .collection import get_model_from_collection_string
+        model = get_model_from_collection_string(collection_string)
+        try:
+            query = model.objects.get_full_queryset()
+        except AttributeError:
+            query = model.objects
```
Comment (normanjaeckel) [inline openslides/utils/cache.py]: Please add a short comment here why we go into except block.

File: openslides/utils/cache.py [hunk 5]
```diff
@@ -203,9 +204,152 @@ def save_data(self, data: UserCacheDataType) -> None:
...
+            # Save the dict into a redis map, if there is at least one value
+            pipe.hmset(
+                self.get_cache_key(collection_string),
+                mapping)
+
+        pipe.execute()
+
+    def add_element(self, collection_string: str, id: int, data: Dict[str, Any]) -> None:
+        """
+        Adds one element to the cache. If the cache does not exists for the user,
```
Comment (normanjaeckel) [inline openslides/utils/cache.py]: user? you mean collection? or collection element?

File: openslides/utils/cache.py [hunk 6]
```diff
@@ -203,9 +204,152 @@ def save_data(self, data: UserCacheDataType) -> None:
...
+    def exists_for_collection(self, collection_string: str) -> bool:
+        return False
+
+    def get_data(self, collection_string: str) -> List[Dict[str, Any]]:
+        from .collection import get_model_from_collection_string
+        model = get_model_from_collection_string(collection_string)
+        try:
+            query = model.objects.get_full_queryset()
+        except AttributeError:
+            query = model.objects
```
Comment (normanjaeckel) [inline openslides/utils/cache.py]: Here the same.

File: openslides/utils/cache.py [hunk 7]
```diff
@@ -203,9 +204,152 @@ def save_data(self, data: UserCacheDataType) -> None:
...
+        return [model.get_access_permissions().get_full_data(instance)
+                for instance in query.all()]
+
+    def get_element(self, collection_string: str, id: int) -> Dict[str, Any]:
+        from .collection import get_model_from_collection_string
+        model = get_model_from_collection_string(collection_string)
+        try:
+            query = model.objects.get_full_queryset()
+        except AttributeError:
+            query = model.objects
```
Comment (normanjaeckel) [inline openslides/utils/cache.py]: Here the same.

**Comment by normanjaeckel** [comment]
Please add CHANGELOG entry.

**Comment by ostcar** [comment]
all done

**Comment by normanjaeckel** [comment]
Ready to test, rebase and merge.

**Pull request merged**